src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/UIHintAttribute.cs

```diff
@@ -159,21 +154,20 @@ public override bool Equals([NotNullWhen(true)] object? obj)
             /// <returns>
             ///     Dictionary of control parameters.
             /// </returns>
-            private IDictionary<string, object?> BuildControlParametersDictionary()
+            private Dictionary<string, object?> BuildControlParametersDictionary()
             {
-                IDictionary<string, object?> controlParameters = new Dictionary<string, object?>();
-
                 object?[]? inputControlParameters = _inputControlParameters;
 
                 if (inputControlParameters == null || inputControlParameters.Length == 0)
                 {
-                    return controlParameters;
+                    return new Dictionary<string, object?>();
                 }
                 if (inputControlParameters.Length % 2 != 0)
                 {
                     throw new InvalidOperationException(SR.UIHintImplementation_NeedEvenNumberOfControlParameters);
                 }
 
+                Dictionary<string, object?> controlParameters = new Dictionary<string, object?>(inputControlParameters.Length / 2);
```

Comment:
Why is this size a good estimate?

Reply:
It's not an estimate. The **inputControlParameters** contains two entries for each kvp that we'll add to the dictionary.
We'll indeed allocate more if a key is null, not a string or a duplicate.
I assumed that it would be ok to focus on the happy path, but (I should not have assumed anything and) we do not gain much from this.
Let me know if you want me to revert this part.

Reply:
> I assumed that it would be ok to focus on the happy path

I've never actually used this type.  Is that the 99.9% case?  Can you point to examples of it being used?  I'm curious how this type shows up in your own code that led you to submit this PR?

Reply:
To be honest, I have never use this type. I just follow all commits in this repo, and I saw another commit that touched this assembly and left room for improvement. This lead me to preparing some other improvements (of which this is just one, others to follow).

I was perhaps too optimistic about the happy path being the common path, I'll just revert this part of the PR.

Reply:
Thanks. If the code isn't hot path, I'd actually prefer we not try to remove the simple LINQ usage by creating open-coded replacements for it.  Or are there other reasons for removing the LINQ dependency?  Is there some usage somewhere of this type that highlights why this is valuable?

Reply:
It was just one of the few uses of Linq in this assembly, and I just thought it may be a good idea to eliminate the use of Linq altogether here (for both perf and allocations). I'll just close the PR then.